### PR TITLE
[rush] Add a --skip-install flag which only updates the shrinkwrap

### PR DESCRIPTION
**Pull request by nickpape-msft**
Sometimes there is a need to update a shrinkwrap file, but you aren't interested in doing a local install (as it may take a lot of time) [e.g. if you are merging from master into your branch.]

### CODE REVIEW

File: apps/rush-lib/src/api/RushLastInstallFlag.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { LastInstallFlag } from './LastInstallFlag';
+import { RushConfiguration } from './RushConfiguration';
+
+export class RushLastInstallFlag extends LastInstallFlag {
```
Comment (pgonzal) [inline apps/rush-lib/src/api/RushLastInstallFlag.ts]: >RushLastInstallFlag [](start = 13, length = 19)

Having a helper makes sense.  Modeling the helper as a subclass (the most brittle form of relationship) doesn't seem best.

File: apps/rush-lib/src/logic/InstallManager.ts [hunk 2]
```diff
@@ -239,15 +242,24 @@ export class InstallManager {
...
               if (!options.noLink) {
-                const linkManager: BaseLinkManager = LinkManagerFactory.getLinkManager(this._rushConfiguration);
                 return linkManager.createSymlinksForProjects(false);
               } else {
-                console.log(os.EOL
-                  + colors.yellow('Since "--no-link" was specified, you will need to run "rush link" manually.'));
+                // since our local links are out of date, go ahead and remove the rush-link.json
+                linkManager.deleteRushLinkJson();
+
+                if (options.skipInstall) {
```
Comment (pgonzal) [inline apps/rush-lib/src/logic/InstallManager.ts]: >skipInstall [](start = 28, length = 11)

In your testing you said that `--skip-install` is not actually faster.  Let's table this until [https://github.com/pnpm/pnpm/pull/1177](https://github.com/pnpm/pnpm/pull/1177) is sorted out.

File: apps/rush-lib/src/api/LastInstallFlag.ts [hunk 3]
```diff
@@ -20,7 +20,7 @@ export class LastInstallFlag {
    * @param folderPath - the folder that this flag is managing
    * @param state - optional, the state that should be managed or compared
   */
-  constructor(folderPath: string, state: Object = {}) {
```
Comment (iclanton) [inline apps/rush-lib/src/api/LastInstallFlag.ts]: >constructor [](start = 2, length = 11)

I thought we had a TSLint rule about this.

File: apps/rush-lib/src/api/RushLastInstallFlag.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  constructor(rushConfiguration: RushConfiguration) {
```
Comment (iclanton) [inline apps/rush-lib/src/api/RushLastInstallFlag.ts]: >constructor [](start = 2, length = 11)

`public `?

File: apps/rush-lib/src/api/RushLastInstallFlag.ts (hunk 1 above)
Comment (iclanton) [reply apps/rush-lib/src/api/RushLastInstallFlag.ts]: Might be better/easier to make this a static initializer.

---
In reply to: [208431171](https://github.com/Microsoft/web-build-tools/pull/766#discussion_r208431171) [](ancestors = 208431171)

File: apps/rush-lib/src/logic/base/BaseLinkManager.ts [hunk 5]
```diff
@@ -157,12 +168,20 @@ export abstract class BaseLinkManager {
       }
     }
 
+    // ensure that the last-install.flag is properly set
+    const lastInstallFlag: RushLastInstallFlag = new RushLastInstallFlag(this._rushConfiguration);
+    if (!lastInstallFlag.isValid()) {
+      return Promise.reject(
+        new Error('Cannot perform "rush link" because the last install flag is missing.'
+          + ' Did you forget to run "rush install"?'));
```
Comment (iclanton) [inline apps/rush-lib/src/logic/base/BaseLinkManager.ts]: >Did you forget to run "rush install"? [](start = 14, length = 37)

I would phrase this as 'Did you run "rush install'?"

Sounds a little less accusatory.

File: apps/rush-lib/src/api/RushLastInstallFlag.ts (hunk 1 above)
Comment (iclanton) [inline apps/rush-lib/src/api/RushLastInstallFlag.ts]: >RushLastInstallFlag [](start = 13, length = 19)

Can you add a comment describing how this is different from `LastInstallFlag`?

**Comment by nickpape-msft** [comment]
Due to this discussion it seems like this flag may not be very useful for us:

https://github.com/pnpm/pnpm/pull/1177

**Comment by pgonzal** [comment]
We're abandoning this because `--skip-install` doesn't work any more due to NPM protocol limitations. There's not much PNPM can do about this.

**Pull request closed without merging**